### ISSUE TITLE
Belgian prices are not fetched because of different resolutions in api response

### ISSUE THREAD

**Issue by Roeland54** [JaccoR/hass-entso-e#195]
So since the api changes of 10/04 the response for the Belgian zone contains PT15M (15 min) resolution data mixed with PT60M (60 min) data. The current parsing logic only parses PT60M data because that was al that was needed until now.

It is currently unclear to me what the solution here should be. We could change the parsing logic to also parse PT15M data. The issue with that is that the German zone also returns PT15M data we explicitly don't want to parse. We could program in some logic to only parse PT15M when the Belgian data is requested but that feels like a 'hack'. 

![image](https://github.com/user-attachments/assets/78870800-db82-4c18-a749-7df8e79ab380)
![image](https://github.com/user-attachments/assets/48347e03-3a0a-4590-a60b-8c31569f54ab)

**Comment by Pluimvee** [comment]
We can either skip the 2nd, 3rd, 4th entry to start with.
In a second we can average the 2nd, 3rd, 4th prices to get back to hour prices

Changing the complete integration to support 15min intervals is quite a large change

**Comment by fboerman** [comment]
dear all, entsoe-py maintainer dropping in. Sorry to hear you had to build your own implementation due to pandas issue.
please note that the SDAC is currently on hourly basis and the 15 min data from Belgium is a bug. For my own datalake I have added a hardcoded exception that resamples the belgium prices.

for the future it is important to note that the SDAC will go to 15 min resolution somewhere Q1 or Q2 2025. This means that you will need the 15 min parser eventually since ALL countries will swap their day ahead prices to that resolution at the same time.

**Comment by wonko** [comment]
this might fix this (for now, not ideal as I don't really care about _real_ 15m resolutions

https://github.com/JaccoR/hass-entso-e/pull/196

**Comment by micvdh** [comment]
Proposing temporary workaround with the change below to api_client.py for mixed resolution in API response : 

    def query_day_ahead_prices(
        self, country_code: Union[Area, str], start: datetime, end: datetime
    ) -> str:
        """
        Parameters
        ----------
        country_code : Area|str
        start : datetime
        end : datetime

        Returns
        -------
        str
        """
        area = Area[country_code.upper()]
        params = {
            "documentType": "A44",
            "in_Domain": area.code,
            "out_Domain": area.code,
        }
        response = self._base_request(params=params, start=start, end=end)

        if response.status_code == 200:
            try:
                root = self._remove_namespace(ET.fromstring(response.content))
                _LOGGER.debug(f"content: {root}")
                series = {}

                # Extract TimeSeries data
                for timeseries in root.findall(".//TimeSeries"):
                    for period in timeseries.findall(".//Period"):
                        resolution = period.find(".//resolution").text

                        _LOGGER.debug(f"Resolution {resolution}")
                        if (resolution != "PT60M" and resolution != "PT30M" and resolution != "PT15M"):
                            continue

                        response_start = period.find(".//timeInterval/start").text
                        start_time = (
                            datetime.strptime(response_start, "%Y-%m-%dT%H:%MZ")
                            .replace(tzinfo=pytz.UTC)
                            .astimezone()
                        )

                        response_end = period.find(".//timeInterval/end").text
                        end_time = (
                            datetime.strptime(response_end, "%Y-%m-%dT%H:%MZ")
                            .replace(tzinfo=pytz.UTC)
                            .astimezone()
                        )

                        _LOGGER.debug(f"Period found is from {start_time} till {end_time}")

                        for point in period.findall(".//Point"):
                            position = point.find(".//position").text
                            price = point.find(".//price.amount").text
                            if resolution == "PT60M":
                                hour = int(position) - 1
                            elif resolution == "PT30M":
                                hour = ((int(position) - 1) / 2)
                            elif resolution == "PT15M":
                                hour = ((int(position) - 1) / 4)
                            else:
                                continue

                            try:
                                series[start_time + timedelta(hours=hour)] = float(price)
                            except:
                                continue

                        # Now fill in any missing hours 
                        current_time = start_time
                        last_price = series[current_time]

                        while current_time < end_time:  # upto excluding! the endtime
                            if current_time in series:
                                last_price = series[current_time]  # Update to the current price
                            else:
                                _LOGGER.debug(f"Extending the price {last_price} of the previous hour to {current_time}")
                                series[current_time] = last_price  # Fill with the last known price
                            current_time += timedelta(hours=1)

                return dict(sorted(series.items()))

            except Exception as exc:
                _LOGGER.debug(f"Failed to parse response content:{response.content}")
                raise exc
        else:
            print(f"Failed to retrieve data: {response.status_code}")
            return None

**Comment by fboerman** [comment]
If it helps I was thinking about making a small library only pulling in prices with no pandas dependency for projects such as these. Would that help?

⁣Get BlueMail for Android ​

**Comment by Roeland54** [comment]
> dear all, entsoe-py maintainer dropping in. Sorry to hear you had to build your own implementation due to pandas issue. please note that the SDAC is currently on hourly basis and the 15 min data from Belgium is a bug. For my own datalake I have added a hardcoded exception that resamples the belgium prices.
> 
> for the future it is important to note that the SDAC will go to 15 min resolution somewhere Q1 or Q2 2025. This means that you will need the 15 min parser eventually since ALL countries will swap their day ahead prices to that resolution at the same time.

I was wondering if you know on how the SDAC and EXAA will be distinguishable from each other when the SDAC changes to 15 min data. 

> If it helps I was thinking about making a small library only pulling in prices with no pandas dependency for projects such as these. Would that help? ⁣

I think that a lightweight low dependency api library could be useful. It would definitely be easier to wait for a package update when entso-e breaks the api again than fixing and maintaining it in this project. In a perfect world that library could then be the core of a library that provides pandas objects. That would eliminate maintenance of redundant functionality. I just don't know how feasible that all is. 

Belgian data for today is in the correct format. So maybe it is fixed. If prices of tomorrow break it again I will look into merging the patch @wonko

**Comment by wonko** [comment]
> 
> Belgian data for today is in the correct format. So maybe it is fixed. If prices of tomorrow break it again I will look into merging the patch @wonko

Keep in mind that the response sent was valid according to the specs as set out in the XML Schema guide. The patch allows to accept these valid responses, doesn't break anything (and even adds tests on the XML, which can be easily extended with other cases in the future). Until there's a uniform solution through some external (or internal) package which deals with everything, it only adds robustness.

But, your call in the end.

**Comment by Roeland54** [comment]
Yes I agree and i like the tests. And have the intention on merging the PR. But this just takes the pressure off releasing a rushed version.

**Comment by Pluimvee** [comment]
PR #202 is now tested and should resolve this issue

**Comment by straybiker** [comment]
Belgian prices work fine here
